### PR TITLE
fix: apply updated schema on refresh

### LINKED ISSUE #80: Schema not updated when calling `refresh`
I've found that calling `server.graphqlGateway.refresh()` doesn't update the application schema. From investigation the federated services are successfully called and their respective schemas built but the refresh function doesn't call `app.graphql.replaceSchema(schema)`

I'll try to submit a PR shortly.

### PR DESCRIPTION
**Pull request by luke88jones**
fixes #80 

- update the gateway refresh function to call `replaceSchema` on the GQL instance

### CODE REVIEW

File: lib/gateway/build-gateway.js [hunk 1]
```diff
@@ -614,6 +614,9 @@ async function buildGateway (serviceMap, gatewayOpts, app, lruGatewayResolvers)
 
       defineResolvers(schema, typeToServiceMap, serviceMap, typeFieldsToService, factory, lruGatewayResolvers)
 
+      this.schema = schema
```
Comment (simoneb) [inline lib/gateway/build-gateway.js:617]: why do you need to set an instance field with the value of the schema?

File: test/schemaRefresh.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            fields {
+              name
+            }
+          }
+        }
+      `
+    })
+  })
+
+  t.same(JSON.parse(res.body), {
```
Comment (simoneb) [inline test/schemaRefresh.js:97]: IIRC you can simply do res.json()

File: test/schemaRefresh.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        name: String!
+        lastName: String!
+      }
+    `)
+  )
+  userService.graphql.defineResolvers(resolvers)
+
+  for (let i = 0; i < 10; i++) {
+    await t.context.clock.tickAsync(200)
+  }
```
Comment (simoneb) [inline test/schemaRefresh.js:131]: why are you ticking the clock in a loop?

File: test/schemaRefresh.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+  // We need the event loop to actually spin twice to
+  // be able to propagate the change
+  await setImmediate()
+  await setImmediate()
```
Comment (simoneb) [inline test/schemaRefresh.js:136]: do you know why?

File: test/schemaRefresh.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    shouldAdvanceTime: true,
+    advanceTimeDelta: 40
+  })
+})
+
+t.afterEach(({ context }) => {
+  context.clock.uninstall()
+})
+
+test('Refreshing gateway schema', async (t) => {
```
Comment (simoneb) [inline test/schemaRefresh.js:25]: I assume this is already the case, but can you please make extra sure that this test fails without the change in the production code?

File: test/schemaRefresh.js (hunk 2 above)
Comment (luke88jones) [reply test/schemaRefresh.js:97]: TBH for this and the other timer issues I was following the setup of the existing tests for refreshing the schema with polling. I'll try removing/update everything

File: test/schemaRefresh.js (hunk 5 above)
Comment (luke88jones) [reply test/schemaRefresh.js:25]: Yes I wrote the test before making the fix to ensure it failed.

File: test/schemaRefresh.js (hunk 4 above)
Comment (luke88jones) [reply test/schemaRefresh.js]: As mentioned this was from following the polling tests. I've now removed these.

File: test/schemaRefresh.js (hunk 3 above)
Comment (luke88jones) [reply test/schemaRefresh.js]: As mentioned this was from following the polling tests. I've now removed these.

**Review by simoneb — APPROVED** [review]
LGTM but I'd like to wait for another review from @codeflyer

File: lib/gateway/build-gateway.js (hunk 1 above)
Comment (luke88jones) [reply lib/gateway/build-gateway.js:617]: Correct 👍

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by simoneb** [comment]
@luke88jones I'm seeing a consistent failure on Node 18 due to a timeout. Because this hasn't happened in the past and because the new behavior you introduced may have interactions with the event loop, I'm tempted to think that this is a legitimate failure, although I'm not sure why it would happen just on Node 18

**Review by marco-ippolito — APPROVED** [review]

**Comment by luke88jones** [comment]
@simoneb Sorry for the silence on this I've been on paternity leave. 

It looks like one of the polling interval tests is timing out. I'll see what I can find.

**Comment by luke88jones** [comment]
@simoneb I've just checked out the main branch locally I have the same failure. I'm trying to track down the issue but it looks  to be something to do with closing the `userService` in this test in the `tests/pollingInterval.js` file 
"Polling schemas (if service is down, schema shouldn't be changed)"

**Pull request merged**